### PR TITLE
feat(AMM)!: make amm.addPool() require minimum collateral

### LINKED ISSUE #4643: Make `amm.addPool` require initial assets
## What is the Problem Being Solved?

Manage production of AMM pools. There needs to be a cost to it to prevent spamming and to ensure initial collateral.

## Description of the Design

* Move `addPool` to a separate contract invitation that charges a fee and requires minimum initial collateral to create the pool
* The cost may be initial liquidity, which gets added to the Reserve.
* The initial cost and minimum liquidity need to be governance parameters.

## Security Considerations

Right now it's too easy and cheap to create pools.

### PR DESCRIPTION
**Pull request by Chris-Hibbert**
closes: #4643

## Description

When creating a new pool for the AMM, there will be a minimum amount
of IST. The caller is expected to provide an equivalent value of the
collateral to balance the AMM pool. The liquidity tokens corresponding
to that minimum will be added to the AssetReserve. If the caller adds
more liquidity than the minimum, they get the extra liquidity tokens.

This required updating a number of tests to initialize their AMMs with
a minimal amount.

### Security Considerations

This adds to the economic security of the INTER protocol, but doesn't have much to do with system security.

### Documentation Considerations

`MinInitialPoolLiquidity` needs to be added to the doc for governance parameters, and the contract description.

### Testing Considerations

updated existing tests.

### CODE REVIEW

**Comment by Chris-Hibbert** [comment]
@turadg  this is now ready for review. #5397 should address the failing test.

File: packages/run-protocol/src/vpool-xyk-amm/AttackersGuide.md
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/AttackersGuide.md:21]: 👍 updating related docs

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/addPool.js]: param types, s'il vous plaît.

from the function signature `isInSecondaries` looks like a boolean

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/addPool.js:36]: for my learning: is this giving earlier feedback or checking something that wouldn't be otherwise? (I assume `makeZCFMint` won't let you override a keyword.)

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/addPool.js:42]: magic number

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/addPool.js]: something still awry with `singlePool.getPriceForInput` type differing between `PoolFacets` and `ReturnType<makePool>`. Which one is correct? If it's `PoolFacets` let's move this cast over to `makePool`'s signature. If it's `makePool`, let's define `PoolFacets` using it.

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/addPool.js]: low risk but this value name reads like it has tokens. consider,
```suggestion
      const centralSurplus =
```

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/addPool.js]: because the `zcf` is untyped. I think you just need `ZCF<{issuers: Issuer[]}>

(also reminder to prefer `ts-expect-error` to `ts-ignore`)

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/addPool.js:153]: thanks for explaining this. any risk in something failing before that happens? (not atomic)

File: packages/run-protocol/src/vpool-xyk-amm/multipoolMarketMaker.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/multipoolMarketMaker.js]: is this before MN-1? should have a ticket so we track it to release

File: packages/run-protocol/src/vpool-xyk-amm/pool.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js:115]: can keep this, no?
```suggestion
  /** @type {import('@agoric/vat-data/src/types').PlusContext<MethodContext, AddLiquidityActual>} */
    addLiquidityActual: (
```

File: packages/run-protocol/src/vpool-xyk-amm/pool.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js]: without an annotation it doesn't know that `feeSeat` is optional

File: packages/run-protocol/src/vpool-xyk-amm/types.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/types.js:95]: the `Invitation` can be typed with the offer args and result. I don't know it offhand but I'd be happy to help.

**Review by turadg — COMMENTED** [review]
Took a pass through `src/`. Haven't reviewed `test/` yet but have to get to a meeting.

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/addPool.js:42]: Do we have a canonical version of this number?  Can I let it default?

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/addPool.js]: done

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/addPool.js:36]: It's an early check so we don't make the mint if `saveIssuer` will fail.

File: packages/run-protocol/src/vpool-xyk-amm/pool.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js:115]: done

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/addPool.js]: Fixed?   Now the bad magic is concentrated around `helper` in `pool.js`.  help?

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/addPool.js]: how's `centralAboveMinimum`?

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/addPool.js]: I updated my template to generate `ts-expect-error`. 

It's still unhappy. I think it's because `find` returns ` '[string, Issuer ] | undefined' `.

File: packages/run-protocol/src/vpool-xyk-amm/addPool.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/addPool.js:153]: It's a valid question, but I think I've covered all the cases.

File: packages/run-protocol/src/vpool-xyk-amm/pool.js
Comment (turadg) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js:115]: I pushed some other type fixes in a71f732d0

File: packages/run-protocol/src/vpool-xyk-amm/pool.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/pool.js:124]: this is a work-around for the cyclic typedef error that happens when a method destructures a facet that it's part of defining. that's a bit of a gotcha. thoughts on how/where to raise awareness?

File: packages/run-protocol/src/vpool-xyk-amm/types.js
Comment (turadg) [inline packages/run-protocol/src/vpool-xyk-amm/types.js]: 👍 

I think the VirtualPool types will be all sorted out with 856aef

File: packages/run-protocol/src/vpool-xyk-amm/types.js
Comment (turadg) [reply packages/run-protocol/src/vpool-xyk-amm/types.js:95]: nevermind, I tried a bit and it's more effort than I thought.

File: packages/run-protocol/test/amm/vpool-xyk-amm/test-liquidity.js
Comment (turadg) [inline packages/run-protocol/test/amm/vpool-xyk-amm/test-liquidity.js:240]: +1 to regular arithmetic in tests

File: packages/run-protocol/test/amm/vpool-xyk-amm/test-xyk-amm-swap.js
Comment (turadg) [inline packages/run-protocol/test/amm/vpool-xyk-amm/test-xyk-amm-swap.js]: solved by 1a7c0a5

File: packages/run-protocol/test/amm/vpool-xyk-amm/test-xyk-amm-swap.js
Comment (turadg) [inline packages/run-protocol/test/amm/vpool-xyk-amm/test-xyk-amm-swap.js]: 💀

File: packages/run-protocol/test/reserve/test-reserve.js
Comment (turadg) [inline packages/run-protocol/test/reserve/test-reserve.js:18]: import from `test/supports.js`

**Review by turadg — APPROVED** [review]
Code LGTM. I pushed a few types fix commits.

I see the tests aren't passing yet but assume it's a matter of tweaking. Please re-request review if more substantial changes are pushed.

File: packages/run-protocol/src/vpool-xyk-amm/multipoolMarketMaker.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/multipoolMarketMaker.js]: #5408

File: packages/run-protocol/src/vpool-xyk-amm/pool.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js:124]: I don't have an idea other than spreading it through code reviews. I get bitten by this while trying to make the types work, and understanding these common workarounds will help me get past that.

File: packages/run-protocol/src/vpool-xyk-amm/pool.js
Comment (Chris-Hibbert) [reply packages/run-protocol/src/vpool-xyk-amm/pool.js:115]: thanks

File: packages/run-protocol/test/amm/vpool-xyk-amm/test-xyk-amm-swap.js
Comment (Chris-Hibbert) [reply packages/run-protocol/test/amm/vpool-xyk-amm/test-xyk-amm-swap.js]: uncommented.  This is valid at this point in the test .

**Comment by Chris-Hibbert** [comment]
Blocked by #5397 

[corrected]

**Comment by Chris-Hibbert** [comment]
I made relatively straightforward changes to integrate with #5397.  I plan to merge this morning when everything is green.

**Comment by mhofman** [comment]
@mergifyio refresh

**Pull request merged**
